### PR TITLE
Rewrite yaml api tests for where-to-build

### PR DESCRIPTION
**Pull request by timsnyder-siv**
@abejgonzalez, I started rewriting the YAML API tests and have gotten far enough that I should open this for others to see.

There are pros and cons of doing what I'm doing.  Here are the ones I see so far:
### PROS
* It is much easier to tightly test for something specific.  For example, your tests just check for non-zero exit code and I think most of them are currently 'passing' in your implementation but they are exiting non-zero because your test recipes are out of date with the YAML API (need to be updated with `default-build-farm` in config_build.yaml) but nothing in your tests notices that the test input is no longer valid.  See https://github.com/firesim/firesim/runs/5461951565?check_suite_focus=true#step:3:47
* Stands alone and is trivial to run locally for testing.  I had trouble trying to run your GH action script locally to see how the tests were expected to behave before my changes.
* Easy to mock out the AWS interaction so that the tests can't actually create AWS resources
* I try to base the tests on the sample config files in deploy/backup-sample-configs and then programmatically modify them so that as we update those, the tests primarily come along for the ride (unless they are testing the specific things that changed).  

### CONS
* actually running fabric stuff under pytest is interesting because pytest blocks all reading from stdin and fabric seems to be hitting the thing pytest installs to prevent reading from stdin during tests.  This is probably just a fabric configuration issue during testing and we shouldn't need to actually execute anything in the fabric tasks to test the YAML api
    * fab-classic seems to use nose instead of pytest.  I've steered away from nose in the past because it is more heavyweight and seemed like it was becoming unmaintained and less popular.
* requires learning about pytest and how fixtures work.  There's about 200 lines of testing infrastructure code for another 200 lines of tests in what I wrote.

#### Related PRs / Issues

Rewrites part of #850 that adds yaml API tests

#### Verilog / AGFI Compatibility

No change in Verilog or AGFI

### Contributor Checklist
- [X] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [X] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [X] Did you add at least one test demonstrating the PR?
- [X] Did you delete any extraneous prints/debugging code?
- [X] Did you state the UI / API impact?
- [X] Did you specify the Verilog / AGFI compatibility impact?

- [X] If applicable, did you regenerate and publicly share default AGFIs?

- [X] If applicable, did you apply the `ci:fpga-deploy` label?

- [X] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

File: deploy/firesim
Comment (abejgonzalez) [inline deploy/firesim:31]: Why is this preferred to using `globals()`?

Comment (abejgonzalez) [reply deploy/firesim:31]: Actually, I see why this is needed, we need this to call the function in the `pytests`. Is there a reason to split this up into `BUILD_` and `RUN_` or is that mainly for organization?

File: deploy/tests/test_yaml_api.py
Comment (abejgonzalez) [inline deploy/tests/test_yaml_api.py:214]: What defines this `buildafi_mock` (fixture right)? This is just removing the calls to the AWS EC2 API right?

File: deploy/tests/test_yaml_api.py
Comment (abejgonzalez) [inline deploy/tests/test_yaml_api.py:247]: Why do we care about the programmatic version of this test? Isn't the main/only API to modify the `builds` section through the file (which is tested in the  `test_invalid_build_section` test?

File: deploy/tests/test_yaml_api.py
Comment (abejgonzalez) [inline deploy/tests/test_yaml_api.py:256]: I wanted to see if the AWS EC2 API would error out saying invalid instance type. We can probably remove this test since it is just testing if the AWS EC2 API can properly catch its own errors.

**Review by abejgonzalez — APPROVED** [review]
This looks much much better than the hacky tests previously done. I see that throughout the code you can set individual yaml sections to ` = None` instead of having to write out yaml directly which is nice!

**Comment by abejgonzalez** [comment]
> * actually running fabric stuff under pytest is interesting because pytest blocks all reading from stdin and fabric seems to be hitting the thing pytest installs to prevent reading from stdin during tests.  This is probably just a fabric configuration issue during testing and we shouldn't need to actually execute anything in the fabric tasks to test the YAML api
>   
>   * fab-classic seems to use nose instead of pytest.  I've steered away from nose in the past because it is more heavyweight and seemed like it was becoming unmaintained and less popular.

Does this manifest in an error?

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:31]: They are split because if `args.task` is in `BUILD_TASKS` then you make a `BuildConfigFile(args)` and pass that into calling the task.  Likewise for `RUN_TASKS` you make a `RuntimeConfig(args)` and pass that into the task.

File: deploy/tests/test_yaml_api.py
Comment (timsnyder-siv) [reply deploy/tests/test_yaml_api.py:214]: The `@patch` decorator passes it in as the first fixture.  I kinda dislike writing it this way and I'm remembering that I dislike it because it's not obvious that `@patch` is doing that.  I'll look at how to write this a better way.

File: deploy/tests/test_yaml_api.py
Comment (timsnyder-siv) [reply deploy/tests/test_yaml_api.py:247]: I only did it to show that you can do it this way for testing purposes.  I suppose I should point out that while you're programmatically modifying the YAML, it gets serialized out to a temporary file that gets read in by firesim.

Comment (abejgonzalez) [reply deploy/tests/test_yaml_api.py:247]: Is it necessary for a unique test for this? Aren't the latter tests doing this anyways?

File: deploy/tests/test_yaml_api.py
Comment (timsnyder-siv) [reply deploy/tests/test_yaml_api.py:214]: Ok.  I went back and reviewed some things.  I think the pytest-mock plugin makes this code look much better.  I was going to use it when I added the pytests but was hesitant to add yet another dependency.  However, looking back at https://github.com/pytest-dev/pytest-mock/#why-bother-with-a-plugin I'm really seeing how much nicer it would be.  I'm going to add that quick this afternoon and finish implementing the rest of the tests so we can merge this.

**Comment by timsnyder-siv** [comment]
> Does this manifest in an error?

Yes,  If I hack one of the tests that doesn't currently blow up during reading the config files and let it actually call buildafi, the output looks like this:
```
   @patch('firesim.buildafi')
    def test_invalid_unmanaged_missing_args(self, buildafi_mock, build_yamls, firesim_argparser):
        # @patch modifies the symbol table with our Mock but we're dispatching through
        # a dict and that needs to be patched as well, otherwise our mock won't be called but
        # the real function will be
        #firesim.BUILD_TASKS['buildafi'] = buildafi_mock
        os.environ['FIRESIM_SOURCED'] = '1'
    
        build_yamls.farm.data['local-build-farm']['args']['hosts'] = None
    
        build_yamls.write()
        args = firesim_argparser.parse_args(['buildafi'] + build_yamls.args)
>       firesim.main(args)

test_yaml_api.py:354: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../firesim.py:285: in main
    BUILD_TASKS[args.task](build_config_file)
../firesim.py:118: in buildafi
    execute(replace_rtl, build_config, hosts=['localhost'])
../../../.local/lib/python3.8/site-packages/fabric/tasks.py:377: in execute
    results[host] = _execute(
../../../.local/lib/python3.8/site-packages/fabric/tasks.py:268: in _execute
    return task.run(*args, **kwargs)
../../../.local/lib/python3.8/site-packages/fabric/tasks.py:168: in run
    return self.wrapped(*args, **kwargs)
../buildtools/buildafi.py:49: in replace_rtl
    run(build_config.make_recipe("PLATFORM=f1 replace-rtl"))
../../../.local/lib/python3.8/site-packages/fabric/network.py:700: in host_prompting_wrapper
    return func(*args, **kwargs)
../../../.local/lib/python3.8/site-packages/fabric/operations.py:1068: in run
    return _run_command(
../../../.local/lib/python3.8/site-packages/fabric/operations.py:919: in _run_command
    result_stdout, result_stderr, status = _execute(
../../../.local/lib/python3.8/site-packages/fabric/operations.py:807: in _execute
    worker.raise_if_needed()
../../../.local/lib/python3.8/site-packages/fabric/thread_handling.py:26: in raise_if_needed
    six.reraise(e[0], e[1], e[2])
../../../miniforge3/envs/firesim_oss_220203/lib/python3.8/site-packages/six.py:719: in reraise
    raise value
../../../.local/lib/python3.8/site-packages/fabric/thread_handling.py:13: in wrapper
    callable(*args, **kwargs)
../../../.local/lib/python3.8/site-packages/fabric/io.py:270: in input_loop
    byte = f.read(1)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.capture.DontReadFromInput object at 0x7fbf58c59b20>, args = (1,)

    def read(self, *args):
>       raise OSError(
            "pytest: reading from stdin while output is captured!  Consider using `-s`."
        )
E       OSError: pytest: reading from stdin while output is captured!  Consider using `-s`.

../../../miniforge3/envs/firesim_oss_220203/lib/python3.8/site-packages/_pytest/capture.py:192: OSError
```

My explanation in the meeting was wrong.  I think that fabric.run by default puts a PTY on stdin and the pytest input capturing is being triggered by that.  I see that new Fabric does have https://docs.fabfile.org/en/2.6/api/testing.html that provides some helpers for testing fabric driven code.

I've been thinking that there is a lot of room for unit-tests of the manager that don't need fabric and then like we discussed, the end-to-end testing at a higher level in CI is great.

File: deploy/tests/test_yaml_api.py
Comment (timsnyder-siv) [reply deploy/tests/test_yaml_api.py:256]: Ok.  Thanks for clarifying.  I wasn't sure if you were somehow validating that in the YAML early or testing AWS.  I'll remove that.

File: deploy/tests/test_yaml_api.py
Comment (timsnyder-siv) [reply deploy/tests/test_yaml_api.py:247]: No, I'll defiintely remove it before we merge.  Was only trying to have food for discussion in the PR.

**Review by davidbiancolin — COMMENTED** [review]
Looks great to me. 

Spitballing, I wonder if _register_task should mandate providing the Mock there. That would clean up the patching. This may all be moot if the plugin work  well though.

File: deploy/firesim
Comment (timsnyder-siv) [inline deploy/firesim:298]: Thinking about @abejgonzalez's question about why I have BUILD_TASKS and RUN_TASKS in two separate datastructures, it would probably be nicer if it was a dict of dicts that also stored the Config class that the task wanted, we could just do this:
```
        f = TASKS[args.task]['function']
        conf = TASKS[args.task]['config_class'](args)
        f(conf)
```
And we could tweak managerinit and terminaterunfarm slightly and not need to have special casing  either.  I like it.

File: .github/scripts/yaml-tests/failing-runtime-files/hwdb-invalid-afi/sample_config_hwdb.yaml
Comment (timsnyder-siv) [inline .github/scripts/yaml-tests/failing-runtime-files/hwdb-invalid-afi/sample_config_hwdb.yaml:2]: @abejgonzalez , This test also seems like something that shouldn't be caught by firesim before asking AWS to really do something with the AGFI.  I'm going to leave it out

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim]: @abejgonzalez , I've pushed the rework I did.  You should see that there is only a unified TASKS datastructure now and that register_task is used as a decorator.  I also switched the tests over to using the pytest-mock `mocker` fixture and I think it makes the mock creation look cleaner than using the `@patch` decorator like I was doing before.

File: deploy/tests/test_yaml_api.py
Comment (timsnyder-siv) [reply deploy/tests/test_yaml_api.py]: This is removed now.

File: deploy/tests/test_yaml_api.py
Comment (timsnyder-siv) [reply deploy/tests/test_yaml_api.py]: I added a test to test_awstools.py that shows `moto` will let you create instances with bogus types when using the mocking library.  I might also add a test for whether aws will properly catch something bogus as an example of how to write a test that does real AWS stuff.  Checking for that case isn't horrible since we are allowing something that a user puts in the config file to go directly to AWS without any kind of validation first.

**Comment by timsnyder-siv** [comment]
@abejgonzalez , I think this is ready to merge into where-to-build if you take another look and agree.

**Review by abejgonzalez — APPROVED** [review]
LGTM with small nits

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:49]: I typically put code refs in backtics.  Thanks for catching this.

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:251]: Sure.  I can do that.   I was just trying to see how much or if I got pushback on putting the args into RuntimeConfig.

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:320]: I'll look more closely at that.  This was refactored into a function without changing the call but it would be good to make things consistent.

File: deploy/firesim
Comment (abejgonzalez) [reply deploy/firesim:251]: I think putting args everywhere is fine for now... I'm not opposed to it right now.

File: deploy/firesim
Comment (abejgonzalez) [reply deploy/firesim:49]: This will probably be checked once we generate some document for the docstrings, so its not a big issue right now.

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim]: done

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim]: Thanks for pointing this out.  I changed all uses of `exit()` to `sys.exit()`

**Pull request merged**
